README.md
# REDCap First Responder COVID-19 ETL Engine - ARCHIVED

**This repository has been archived and is no longer actively maintained.**

This project was archived as of October 1, 2025. The project for which it was written is now over. There's no funding to provide further maintenance for other projects. Please don't hesitate to use this code in accordance with the license; however, the authors are unable to offer any additional support.

-----------


This project provides extract, transform, and load (ETL) tools in support of the First Responders COVID-19 Testing project at the University of Florida. The ETL tools are RScripts run by a Docker container.

## Prerequisites

This script use R and these R packages:

    tidyverse
    dotenv
    REDCapR
    openxlsx
    sendmailR

To build the Docker container, you will need only Docker.

This project is designed to move data between two REDCap projects that work together to collect and curate the data in a COVID-19 testing workflow. The source project, referred to as the _results_ project in the code and configuration files, is provided as a REDCap project XML file at [`./examples/First_Responder_COVID19_Results_Upload.xml`](./examples/First_Responder_COVID19_Results_Upload.xml).  The target project, referred to as the _survey_ project in the code and configuration files, is available at as [`First_Responder_COVID19.xml`](https://github.com/ctsit/fr_covidata/blob/master/example/First_Responder_COVID19.xml) in the [fr_covidata REDCap module](https://github.com/ctsit/fr_covidata).

This script uses the REDCap API to move data between the two projects. The API must be enabled on the REDCap project and the host where this script runs will need to have access to it.


## Setup and Configuration

This script is configured entirely via the environment. An example `.env` files are provided as `./example.env` and `./example_pky.env` To use one of these files, copy it to the name `.env` and customize according to your project needs. Follow these steps to build the required components and configure the script's `.env` file.

1. Create each of the REDCap projects from the project XML files. We will refer to these two projects as _survey_ and _results_ for the remainder of this document.
1. In both the survey and results projects, give a user User Rights of _Full Data Set_ for _Data Exports_
1. In each project, that user will need an API key in _each_ project.
1. Add the new API keys to the .env file taking care to not confuse the two keys.
1. Change the `*_PROJECT_TITLE` and `*_PROJECT_PID` fields to match the result and survey projects.
1. Set `TIME_ZONE` to assure that time stamps used in the file names and the email are accurate.
1. Revise the `EMAIL_*` and `SMTP_SERVER` settings to reflect your local needs.


## Running the ETL scripts

Each ETL job in this system is an Rscript run via Docker. The Docker containers are hosted on a Linux host with access to the REDCap API interface and a mail server. At UF, this host is `tools4.ctsi.ufl.edu`. Scripts can also be run at the command line or in RStudio. In each case the script will read its configuration from the `.env` file.

To build the image and run the report using docker within the project directory do:

`docker build -t fr_covidata_engine .`

and run the script using docker with a command something like this:

`docker run --rm --env-file <path_to_dir_full_of_env_files>/fr_dev.env fr_covidata_engine Rscript load_results_into_survey_project.R`

Example cron scripts that could run the containers on a regular basis are provided in [./examples/*.cron](./examples/)


## Release and Deployment

This project uses the Git Flow workflow for releases. Every release should be versioned and have a ChangeLog entry that describes the new features and bug fixes. Every release should also be accompanied by an updated `VERSION` file to allow image builds to be tagged as they are built by the `build.sh`

To deploy a new release on tools4, execute this series of commands or an equivalent from your home directory:

```
git clone https://github.com/ctsit/fr_covidata_engine.git
cd fr_covidata_engine
git pull
sudo ./build.sh
```


## Testing workflows

To test the the `load_results*` scripts, follow these steps:

1. Write the environment file as described above.
1. Create appointment records in the Survey project. 10-15 records make a good test. 
1. Use `load_fake_data_into_upload.R` to generate fake result data in the Results project. It will derive a swab result value from the the research_encounter_id. 
1. Run `load_results_into_pky_projects.R` or `load_results_into_survey_project.R` according to your need.
1. If you are testing the PKY projects with `load_results_into_pky_projects.R`, continue the test by adding appointment records on the next _follow-up_ event in the serial project. 
1. Create fake result data based on these records by running `load_fake_data_from_serial_into_upload.R`.
1. Test the complete workflow by rerunning `load_results_into_pky_projects.R`.


CHANGELOG.md
# Change Log
All notable changes to the REDCap First Responder COVID-19 ETL Engine project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.8.0] - 2021-02-03
### Summary
This release copies the old swab result script and updates the load results
script to handle saliva results upload instead of swab.

### Updated
* Remove swab results from exclusion logic and replace with saliva (Samantha Emerson)
* Replace swab variable with saliva for result processing (Samantha Emerson)

## [0.7.0] - 2020-11-03
### Changed
 * Added .rproject.user to gitignore file (Mel Moreno)
 * Added handling for igm and iga antibodies to the load_fr_igm_iga_results_into_survey_project.R file to handle results between the upload and testing project (Mel Moreno)

## [0.6.0] - 2020-08-06
### Changed
 * Added `research_encounter_id` to Swab Results Imported report (Taeber Rapczak)


## [0.5.1] - 2020-07-02
### Changed
- Allow indeterminate lab results to be loaded for FR IgG (Philip Chase)
- Remove leading zero from minute in cron file (Philip Chase)


## [0.5.0] - 2020-06-16
### Added
- Add Docker image build script (Philip Chase)


## [0.4.0] - 2020-06-16
### Added
- Add FR IgG load script (Philip Chase)
- Add VERSION file referenced in 'Release and Deployment' (Philip Chase)
- Add Release and Deployment section to README (Philip Chase)
- Add prod cron file for FR (Philip Chase)
- Add the URI parameter to example.env (Philip Chase)

### Changed
- Align Dockerfile with tt_engine work (Philip Chase)
- Use URI env var in load fake data (Philip Chase)


## [0.3.0] - 2020-04-24
### Added
- Set form_complete fields in PKY (Philip Chase)


## [0.2.0] - 2020-04-23
### Added
 - Verify the checksums of barcodes before save data to REDCap (Laurence James-Woodley)
 - Add scripts to load data nd test data for PKY (Philip Chase)
 - initialze consecutive negative swab results to zero (Philip Chase)

### Changed
 - Allow one docker image to run any of the scripts (Philip Chase)


## [0.1.0] - 2020-04-14
### Summary
 - First release of fr_covidata_engine
 - Supports swab results: pos, neg, and ina
 - Updates only blank results in survey project
 - Provides a log of upload results and non-loadable for swabs
 - Writes a custom email body using good/bad result count, project titles, PIDs, and URL to result project for address errors
 - Emails log file to configurable addressees


load_results_into_survey_project.R
library(tidyverse)
library(dotenv)
library(REDCapR)
library(openxlsx)
library(sendmailR)

source("functions.R")
# set the timezone
Sys.setenv(TZ = Sys.getenv("TIME_ZONE"))

# echo details from the .env file we read
Sys.getenv("INSTANCE")
Sys.getenv("PROJECT")

script_run_time <- strftime(Sys.time(), format = "%Y%m%d_%H%M")

# email credentials
email_server <- list(smtpServer = Sys.getenv("SMTP_SERVER"))
email_from <- Sys.getenv("EMAIL_FROM")
email_to <- unlist(strsplit(Sys.getenv("EMAIL_TO"), " "))
email_cc <- unlist(strsplit(Sys.getenv("EMAIL_CC"), " "))
email_subject <- paste(Sys.getenv("EMAIL_SUBJECT"), script_run_time)

# NOTE: For production script the data will be read from pid 8270
# to get the research_encounter_id which will then be written to pid 8258

# read data from survey project (prod pid 8258)
survey_project_read <-
  redcap_read_oneshot(
    redcap_uri = "https://redcap.ctsi.ufl.edu/redcap/api/",
    token = Sys.getenv("SURVEY_TOKEN")
  )$data %>%
  filter(!is.na(research_encounter_id)) %>%
  select(
    record_id,
    redcap_event_name,
    research_encounter_id,
    covid_19_swab_result,
    igg_antibodies,
    igm_antibodies
  )

# survey records without swab data
survey_swab_data <- survey_project_read %>%
  filter(is.na(covid_19_swab_result)) %>%
  select(record_id, redcap_event_name, research_encounter_id)

# read data from result upload project, (prod pid 8270)
result_project_read <-
  redcap_read_oneshot(
    redcap_uri = "https://redcap.ctsi.ufl.edu/redcap/api/",
    token = Sys.getenv("RESULT_TOKEN")
  )$data %>%
  rowwise() %>%
  mutate(verified_id = verifyLuhn(record_id))

# Run these two tests on entire dataset so we'll know that the corrections have
# been made

# flag incorrect barcodes
result_id_with_bad_checksum <- result_project_read %>%
  filter(!verified_id) %>%
  mutate(reason_not_imported = "bad checksum in barcode id") %>%
  select(
    record_id,
    covid_19_swab_result,
    verified_id,
    reason_not_imported
  )

# on the low chance that the barcode passes checksum but does not match an id
# in the survey project
result_id_without_match <- result_project_read %>%
  filter(verified_id) %>%
  anti_join(survey_project_read,
    by = c("record_id" = "research_encounter_id")
  ) %>%
  mutate(reason_not_imported = "no match in target project") %>%
  select(
    record_id,
    covid_19_swab_result,
    verified_id,
    reason_not_imported
  )

# make result upload file for swabs
swab_result <- result_project_read %>%
  filter(!is.na(record_id)) %>%
  select(
    research_encounter_id = record_id,
    covid_19_swab_result,
    verified_id
  ) %>%
  filter(!is.na(covid_19_swab_result)) %>%
  # join to get records in survey project without swab results
  inner_join(survey_swab_data, by = c("research_encounter_id")) %>%
  mutate(
    covid_19_swab_result = case_when(
      str_detect(str_to_lower(covid_19_swab_result), "pos") ~ "1",
      str_detect(str_to_lower(covid_19_swab_result), "neg") ~ "0",
      str_detect(str_to_lower(covid_19_swab_result), "ina") ~ "99",
      TRUE ~ covid_19_swab_result
    )
  ) %>%
  select(
    record_id,
    redcap_event_name,
    research_encounter_id,
    covid_19_swab_result,
    verified_id
  ) %>%
  arrange(record_id)

# only send an email when there are new swab results
if (nrow(swab_result) > 0) {
  # create folder to store output
  output_dir <- paste0("fr_covid19_import_log_", script_run_time)
  dir.create(output_dir, recursive = T)

  # write data to survey project
  swab_result_to_import <- swab_result %>%
    filter(covid_19_swab_result %in% c("1", "0", "99") &
      verified_id) %>%
    select(-verified_id)

  bad_swab_result <- swab_result %>%
    select(-redcap_event_name) %>%
    filter(!covid_19_swab_result %in% c("1", "0", "99") &
      verified_id) %>%
    mutate(reason_not_imported = "improper value for swab result") %>%
    mutate_at(vars(record_id), as.character) %>%
    bind_rows(result_id_with_bad_checksum) %>%
    bind_rows(result_id_without_match)

  # only write to redcap when there are legit records
  if (nrow(swab_result_to_import) > 0) {
    # Remove the research_encounter_id field to minimize the data being changed.
    survey_records_to_update <- swab_result_to_import %>%
      select(-research_encounter_id)
    redcap_write_oneshot(
      survey_records_to_update,
      redcap_uri = "https://redcap.ctsi.ufl.edu/redcap/api/",
      token = Sys.getenv("SURVEY_TOKEN")
    )
  }

  all_output <-
    list(
      "Swab Results Imported" = swab_result_to_import,
      "Swab Results Not Imported" = bad_swab_result
    )

  write.xlsx(
    all_output,
    paste0(output_dir, "/swab_result_log_", script_run_time, ".xlsx"),
    na = ""
  )

  # Zip the reports generated
  zipfile_name <- paste0(output_dir, ".zip")
  zip(zipfile_name, output_dir)

  # attach the zip file and email it
  attachment_object <- mime_part(zipfile_name, zipfile_name)
  project_title <- Sys.getenv("SURVEY_PROJECT_TITLE")
  project_pid <- Sys.getenv("SURVEY_PROJECT_PID")
  result_project_pid <- Sys.getenv("RESULT_PROJECT_PID")

  body <-
    paste0(
      "The attached file(s) includes a log of all results that were uploaded",
      " to the REDCap project, ",
      project_title,
      " (PID ",
      project_pid,
      ")",
      "\n\nNumber of records uploaded: ",
      nrow(swab_result_to_import),
      "\nNumber of records not uploaded: ",
      nrow(bad_swab_result),
      "\n\nIf there are records that were not uploaded, then there were",
      " improper values in the swab result column or incorrect record_ids were used",
      " Please review the Swab Results Not Imported tab in the attached log file",
      " then update these records at ",
      "https://redcap.ctsi.ufl.edu/redcap/redcap_v9.3.5/index.php?pid=",
      result_project_pid
    )

  body_with_attachment <- list(body, attachment_object)

  # send the email with the attached output file
  sendmail(
    from = email_from,
    to = email_to,
    cc = email_cc,
    subject = email_subject,
    msg = body_with_attachment,
    control = email_server
  )

  # uncomment to delete output once on tools4
  # unlink(zipfile_name)
}


load_fake_igg_into_pky_timetwo_projects.R
library("tidyverse")

#Results from Julia (emailed)
results<-read.csv("results.csv", header= TRUE)

#From Coronavirus Infection Among Children in North Central Florida - Production
redcap_production<-read.csv("CoronavirusInfection_DATA_2020-07-28_1211.csv", header= TRUE)

#From scheduling project 
scheduling<-read.csv("PKYongeCOVID19Schedu_DATA_2020-07-28_1455.csv", header= TRUE)


redcap_production1<- redcap_production %>% 
  mutate(igg_antibodies = as.character(igg_antibodies)) %>%
  mutate(igm_antibodies = as.character(igm_antibodies)) %>% 
  mutate(research_encounter_id = as.character(research_encounter_id)) %>% 
  select(record_id= ï..record_id, research_encounter_id, redcap_event_name, ce_family_id, test_date_and_time)%>% 
  drop_na(research_encounter_id, ce_family_id)


results1<- results %>% 
  filter(!is.na(record_id)) %>% 
  mutate(igg_antibodies = as.character(igg_antibodies)) %>% 
  mutate(research_encounter_id = as.character(record_id)) %>% 
  mutate(igg_antibodies = case_when(
    str_detect(str_to_lower(igg_antibodies), "pos") ~ "1",
    str_detect(str_to_lower(igg_antibodies), "neg") ~ "0",
    str_detect(str_to_lower(igg_antibodies), "indeterminate") ~ "2",
    TRUE ~ igg_antibodies
  )) %>% 
  select(research_encounter_id, igg_antibodies)
  

scheduling1<- scheduling %>% 
  mutate(igg_antibodies = as.character(igg_antibodies)) %>%
  mutate(igm_antibodies = as.character(igm_antibodies)) %>% 
  mutate(research_encounter_id = as.character(research_encounter_id)) %>%
  select(record_id= ï..record_id, research_encounter_id, test_date_and_time, redcap_event_name, ce_family_id)


##PKY project 

results_pky<- join(results1, scheduling1, by= "research_encounter_id")

pky_results<-results_pky %>% 
  filter(!is.na(test_date_and_time))

write.csv(pky_results, "pky_results.csv")



### Time .2 

library("plyr")

results_dataset<- dplyr::left_join(redcap_production1, results1, by= "research_encounter_id")

timetwo_results<-results_dataset %>% 
  filter(!research_encounter_id =="") %>% 
  filter(!is.na(igg_antibodies)) 

scheduling2<- scheduling %>%
  group_by(ce_family_id) %>% 
  mutate(test_date_and_time = lead(test_date_and_time)) %>% 
select(test_date_and_time, ce_family_id) 
  
results_schedule<- merge(timetwo_results, scheduling2, by= "ce_family_id" )

colnames(results_schedule)<- c("ce_family_id", "record_id", "research_encounter_id",  "redcap_event_name", "remove", "igg_antibodies",  "test_date_and_time")


results_schedule<- results_schedule %>% 
  select("record_id", "research_encounter_id", "redcap_event_name", "ce_family_id", "igg_antibodies",  "test_date_and_time")


results_schedule<- results_schedule[!duplicated(results_schedule$"research_encounter_id"), ] 

write.csv(results_schedule, "results_schedule.csv")

results_test<- results_schedule %>% 
  select(research_encounter_id)

results_total<- results %>% 
  select(research_encounter_id= record_id)


library("arsenal")

summary(comparedf(redcap_production, results_total, by= "research_encounter_id"))

summary(comparedf(pky_results, results_total, by= "research_encounter_id"))


write.csv(results_schedule, "results_schedule.csv")


load_fr_igg_results_into_survey_project.R
library(tidyverse)
library(dotenv)
library(REDCapR)
library(openxlsx)
library(sendmailR)

source("functions.R")
# set the timezone
Sys.setenv(TZ = Sys.getenv("TIME_ZONE"))

# echo details from the .env file we read
Sys.getenv("INSTANCE")
Sys.getenv("PROJECT")

script_run_time <- strftime(Sys.time(), format = "%Y%m%d_%H%M") 

# email credentials
email_server <- list(smtpServer = Sys.getenv("SMTP_SERVER"))
email_from <- Sys.getenv("EMAIL_FROM")
email_to <- unlist(strsplit(Sys.getenv("EMAIL_TO")," "))
email_cc <- unlist(strsplit(Sys.getenv("EMAIL_CC")," "))
email_subject <- paste(Sys.getenv("EMAIL_SUBJECT"), script_run_time)

# NOTE: For production script the data will be read from pid 8270
# to get the research_encounter_id which will then be written to pid 8258

# read data from survey project (prod pid 8258)
survey_project_read <- redcap_read_oneshot(redcap_uri = Sys.getenv("URI"),
                               token = Sys.getenv("SURVEY_TOKEN"))$data %>%
  filter(!is.na(research_encounter_id)) %>%
  select(record_id, redcap_event_name, research_encounter_id, 
         saliva_result, igg_antibodies, igm_antibodies)

# survey records without a lab result
survey_lab_data <- survey_project_read %>%
  filter(is.na(igg_antibodies)) %>%
  select(record_id, redcap_event_name, research_encounter_id)

# read data from result upload project, (prod pid 8270)
result_project_read <- redcap_read_oneshot(redcap_uri = Sys.getenv("URI"),
                               token = Sys.getenv("RESULT_TOKEN"))$data %>% 
  rowwise() %>% 
  mutate(verified_id = verifyLuhn(record_id)) %>%
  mutate(igg_antibodies = as.character(igg_antibodies)) %>%
  mutate(igm_antibodies = as.character(igm_antibodies))


# Run these two tests on entire dataset so we'll know that the corrections have
# been made

# flag incorrect barcodes
result_id_with_bad_checksum <- result_project_read %>% 
  filter(!verified_id) %>% 
  mutate(reason_not_imported = 'bad checksum in barcode id') %>% 
  select(record_id, igg_antibodies, verified_id,
      reason_not_imported)

# on the low chance that the barcode passes checksum but does not match an id
# in the survey project
result_id_with_no_match_in_survey <- result_project_read %>% 
  filter(verified_id) %>% 
  anti_join(survey_project_read, by = c("record_id" = "research_encounter_id")) %>% 
  mutate(reason_not_imported = 'no match in target project') %>% 
  select(record_id, igg_antibodies, verified_id, 
         reason_not_imported)

# make result upload file for this test
lab_result <- result_project_read %>% 
  filter(!is.na(record_id)) %>% 
  select(research_encounter_id = record_id, igg_antibodies, verified_id) %>%
  filter(!is.na(igg_antibodies)) %>%  
  # join to get records in survey project without results
  inner_join(survey_lab_data, by=c("research_encounter_id")) %>%
  mutate(igg_antibodies = case_when(
    str_detect(str_to_lower(igg_antibodies), "pos") ~ "1",
    str_detect(str_to_lower(igg_antibodies), "neg") ~ "0",
    str_detect(str_to_lower(igg_antibodies), "indeterminate") ~ "2",
    str_detect(str_to_lower(igg_antibodies), "inadequate") ~ "99",
    TRUE ~ igg_antibodies
  )) %>%
  select(record_id, redcap_event_name, igg_antibodies, verified_id) %>%
  arrange(record_id) 

# only send an email when there are new results
if (nrow(lab_result) > 0){

  # create folder to store output
  output_dir <- paste0("fr_covid19_import_log_", script_run_time)
  dir.create(output_dir, recursive = T)
  
  # write data to survey project
  lab_result_to_import <- lab_result %>% 
  filter(igg_antibodies %in% c("1","0","2","99") & verified_id) %>% 
    select(-verified_id)
  
  bad_lab_result <- lab_result %>% 
    select(-redcap_event_name) %>% 
    filter(!igg_antibodies %in% c("1","0","2","99") & verified_id) %>% 
    mutate(reason_not_imported = "improper value for result") %>%   
    mutate_at(vars(record_id), as.character) %>% 
    bind_rows(result_id_with_bad_checksum) %>% 
    bind_rows(result_id_with_no_match_in_survey)
  
  # only write to redcap when there are legit records
  if(nrow(lab_result_to_import) > 0 ){
  redcap_write_oneshot(lab_result_to_import,
                       redcap_uri = Sys.getenv("URI"),
                       token = Sys.getenv("SURVEY_TOKEN"))
  }
  
  all_output <- list("Results Imported" = lab_result_to_import,
                     "Results Not Imported" = bad_lab_result)
  
  write.xlsx(all_output, 
             paste0(output_dir, "/lab_result_log_", script_run_time, ".xlsx"), 
             na = "")
  
  # Zip the reports generated
  zipfile_name <-  paste0(output_dir, ".zip")
  zip(zipfile_name, output_dir)
  
  # attach the zip file and email it
  attachment_object <- mime_part(zipfile_name, zipfile_name)
  project_title <- Sys.getenv("SURVEY_PROJECT_TITLE")
  project_pid   <- Sys.getenv("SURVEY_PROJECT_PID")
  result_project_pid <- Sys.getenv("RESULT_PROJECT_PID")
  
  body <- paste0("The attached file(s) includes a log of all results that were uploaded",
               " to the REDCap project, ", project_title, " (PID ", project_pid ,")",
               "\n\nNumber of records uploaded: ", nrow(lab_result_to_import),
               "\nNumber of records not uploaded: ", nrow(bad_lab_result),
               "\n\nIf there are records that were not uploaded, then there were",
               " improper values in the result column or incorrect record_ids were used", 
               " Please review the Results Not Imported tab in the attached log file",
               " then update these records at ",
               "https://redcap.ctsi.ufl.edu/redcap/redcap_v9.3.5/index.php?pid=", result_project_pid)
  
  body_with_attachment <- list(body, attachment_object)
  
  # send the email with the attached output file
  sendmail(from = email_from, to = email_to, cc = email_cc,
           subject = email_subject, msg = body_with_attachment,
           control = email_server)
  
  # uncomment to delete output once on tools4
  # unlink(zipfile_name)
  
} 


load_fr_igm_iga_results_into_survey_project.R
library(tidyverse)
library(dotenv)
library(REDCapR)
library(openxlsx)
library(sendmailR)

source("functions.R")
# set the timezone
Sys.setenv(TZ = Sys.getenv("TIME_ZONE"))

# echo details from the .env file we read
Sys.getenv("INSTANCE")
Sys.getenv("PROJECT")

script_run_time <- strftime(Sys.time(), format = "%Y%m%d_%H%M") 

# email credentials
email_server <- list(smtpServer = Sys.getenv("SMTP_SERVER"))
email_from <- Sys.getenv("EMAIL_FROM")
email_to <- unlist(strsplit(Sys.getenv("EMAIL_TO")," "))
email_cc <- unlist(strsplit(Sys.getenv("EMAIL_CC")," "))
email_subject <- paste(Sys.getenv("EMAIL_SUBJECT"), script_run_time)

# NOTE: For production script the data will be read from pid 8270
# to get the research_encounter_id which will then be written to pid 8258

# read data from survey project (prod pid 8258)
survey_project_read <- redcap_read_oneshot(redcap_uri = Sys.getenv("URI"),
                                           token = Sys.getenv("SURVEY_TOKEN"))$data %>%
  filter(!is.na(research_encounter_id)) %>%
  select(record_id, redcap_event_name, research_encounter_id, 
         covid_19_swab_result, igg_antibodies, igm_antibodies, iga_antibodies)

# survey records without a lab result
survey_lab_data <- survey_project_read %>%
  filter(is.na(igg_antibodies)) %>%
  select(record_id, redcap_event_name, research_encounter_id)

# read data from result upload project, (prod pid 8270)
result_project_read <- redcap_read_oneshot(redcap_uri = Sys.getenv("URI"),
                                           token = Sys.getenv("RESULT_TOKEN"))$data %>% 
  rowwise() %>% 
  mutate(verified_id = verifyLuhn(record_id)) %>%
  mutate(igg_antibodies = as.character(igg_antibodies)) %>%
  mutate(igm_antibodies = as.character(igm_antibodies)) %>% 
  mutate(iga_antibodies = as.character(iga_antibodies)) 


# Run these two tests on entire dataset so we'll know that the corrections have
# been made

# flag incorrect barcodes
result_id_with_bad_checksum <- result_project_read %>% 
  filter(!verified_id) %>% 
  mutate(reason_not_imported = 'bad checksum in barcode id') %>% 
  select(record_id, igg_antibodies, igm_antibodies, iga_antibodies, verified_id,
         reason_not_imported)

# on the low chance that the barcode passes checksum but does not match an id
# in the survey project
result_id_with_no_match_in_survey <- result_project_read %>% 
  filter(verified_id) %>% 
  anti_join(survey_project_read, by = c("record_id" = "research_encounter_id")) %>% 
  mutate(reason_not_imported = 'no match in target project') %>% 
  select(record_id, igg_antibodies, igm_antibodies, iga_antibodies, verified_id, 
         reason_not_imported)

# make result upload file for this test
lab_result <- result_project_read %>% 
  filter(!is.na(record_id)) %>% 
  select(research_encounter_id = record_id, igg_antibodies, igm_antibodies, iga_antibodies, verified_id) %>%
  filter(!is.na(igg_antibodies)) %>%  
  # join to get records in survey project without swab results
  inner_join(survey_lab_data, by=c("research_encounter_id")) %>%
  mutate(igg_antibodies = case_when(
    str_detect(str_to_lower(igg_antibodies), "pos") ~ "1",
    str_detect(str_to_lower(igg_antibodies), "neg") ~ "0",
    str_detect(str_to_lower(igg_antibodies), "indeterminate") ~ "2",
    str_detect(str_to_lower(igg_antibodies), "inadequate") ~ "99",
    TRUE ~ igg_antibodies
  )) %>%
  mutate(igm_antibodies = case_when(
    str_detect(str_to_lower(igm_antibodies), "pos") ~ "1",
    str_detect(str_to_lower(igm_antibodies), "neg") ~ "0",
    str_detect(str_to_lower(igm_antibodies), "indeterminate") ~ "2",
    str_detect(str_to_lower(igm_antibodies), "inadequate") ~ "99",
    TRUE ~ igm_antibodies
  )) %>%
  mutate(iga_antibodies = case_when(
    str_detect(str_to_lower(iga_antibodies), "pos") ~ "1",
    str_detect(str_to_lower(iga_antibodies), "neg") ~ "0",
    str_detect(str_to_lower(iga_antibodies), "indeterminate") ~ "2",
    str_detect(str_to_lower(iga_antibodies), "inadequate") ~ "99",
    TRUE ~ iga_antibodies
  )) %>%
  select(record_id, redcap_event_name, igg_antibodies, igm_antibodies, iga_antibodies, verified_id) %>%
  arrange(record_id) 

# only send an email when there are new swab results
if (nrow(lab_result) > 0){
  
  # create folder to store output
  output_dir <- paste0("fr_covid19_import_log_", script_run_time)
  dir.create(output_dir, recursive = T)
  
  # write data to survey project
  lab_result_to_import <- lab_result %>% 
    filter(igm_antibodies %in% c("1","0","2","99") & verified_id) %>% 
    select(-verified_id)
  
  bad_lab_result <- lab_result %>% 
    select(-redcap_event_name) %>% 
    filter(!igg_antibodies %in% c("1","0","2","99") & verified_id) %>% 
    mutate(reason_not_imported = "improper value for result") %>%   
    mutate_at(vars(record_id), as.character) %>% 
    bind_rows(result_id_with_bad_checksum) %>% 
    bind_rows(result_id_with_no_match_in_survey)
  
   #only write to redcap when there are legit records
  if(nrow(lab_result_to_import) > 0 ){
  redcap_write_oneshot(lab_result_to_import,
                       redcap_uri = Sys.getenv("URI"),
                       token = Sys.getenv("SURVEY_TOKEN"))
   }
  
  all_output <- list("Results Imported" = lab_result_to_import,
                     "Results Not Imported" = bad_lab_result)
  
  write.xlsx(all_output, 
             paste0(output_dir, "/lab_result_log_", script_run_time, ".xlsx"), 
             na = "")
  
  # Zip the reports generated
  zipfile_name <-  paste0(output_dir, ".zip")
  zip(zipfile_name, output_dir)
  
  # attach the zip file and email it
  attachment_object <- mime_part(zipfile_name, zipfile_name)
  project_title <- Sys.getenv("SURVEY_PROJECT_TITLE")
  project_pid   <- Sys.getenv("SURVEY_PROJECT_PID")
  result_project_pid <- Sys.getenv("RESULT_PROJECT_PID")
  
  body <- paste0("The attached file(s) includes a log of all results that were uploaded",
                 " to the REDCap project, ", project_title, " (PID ", project_pid ,")",
                 "\n\nNumber of records uploaded: ", nrow(lab_result_to_import),
                 "\nNumber of records not uploaded: ", nrow(bad_lab_result),
                 "\n\nIf there are records that were not uploaded, then there were",
                 " improper values in the result column or incorrect record_ids were used", 
                 " Please review the Results Not Imported tab in the attached log file",
                 " then update these records at ",
                 "https://redcap.ctsi.ufl.edu/redcap/redcap_v9.3.5/index.php?pid=", result_project_pid)
  
  body_with_attachment <- list(body, attachment_object)
  
  # send the email with the attached output file
  sendmail(from = email_from, to = email_to, cc = email_cc,
           subject = email_subject, msg = body_with_attachment,
           control = email_server)
  
  # uncomment to delete output once on tools4
  # unlink(zipfile_name)
  
} 
